java/piranha/src/main/java/com/uber/piranha/config/PiranhaFlagMethodRecord.java

```diff
@@ -0,0 +1,141 @@
+  private static final String ARGUMENT_INDEX_KEY = "argumentIndex";
+  private static final String RETURN_TYPE_STRING = "returnType";
+  private static final String RECEIVER_TYPE_STRING = "receiverType";
+
+  /**
+   * Holds the mapping of flagType string to API. Eg: "treated" -> API.IS_TREATED. Is initialized
+   * once and then accessed without updating.
+   */
+  private static final ImmutableMap<String, XPFlagCleaner.API> flagTypeToAPIMap =
+      initializeFlagTypeToAPIMap();
+
+  private final String methodName;
+  private final XPFlagCleaner.API apiType;
+  private final Optional<Integer> argumentIdx;
+  private final Optional<String> receiverType;
+  private final Optional<String> returnType;
+
+  PiranhaFlagMethodRecord(
+      String methodName,
+      String flagTypeString,
+      Optional<Integer> argumentIdx,
+      Optional<String> receiverType,
+      Optional<String> returnType) {
+    this.methodName = methodName;
+    this.apiType = flagTypeToAPIMap.getOrDefault(flagTypeString, XPFlagCleaner.API.UNKNOWN);
+    this.argumentIdx = argumentIdx;
+    this.receiverType = receiverType;
+    this.returnType = returnType;
+  }
+
+  @Override
+  public String getMethodName() {
+    return methodName;
+  }
+
+  public XPFlagCleaner.API getApiType() {
+    return apiType;
+  }
+
+  @Override
+  public Optional<Integer> getArgumentIdx() {
+    return argumentIdx;
+  }
+
+  @Override
+  public Optional<String> getReceiverType() {
+    return receiverType;
+  }
+
+  @Override
+  public Optional<String> getReturnType() {
+    return returnType;
+  }
+
+  private static ImmutableMap<String, XPFlagCleaner.API> initializeFlagTypeToAPIMap() {
+    ImmutableMap.Builder<String, XPFlagCleaner.API> builder = new ImmutableMap.Builder<>();
+    builder.put("treated", XPFlagCleaner.API.IS_TREATED);
+    builder.put("control", XPFlagCleaner.API.IS_CONTROL);
+    builder.put("set_treated", XPFlagCleaner.API.SET_TREATED);
+    builder.put("set_control", XPFlagCleaner.API.SET_CONTROL);
+    builder.put("empty", XPFlagCleaner.API.DELETE_METHOD);
+    builder.put("treatmentGroup", XPFlagCleaner.API.IS_TREATMENT_GROUP_CHECK);
+    return builder.build();
+  }
+
+  /**
+   * Parse the entry for a single method from piranha.json that has been previously decoded into a
+   * map
+   *
+   * @param methodPropertyEntry The decoded json entry (as a Map of property names to values)
+   * @param isArgumentIndexOptional Whether argumentIdx should be treated as optional
+   * @return A PiranhaFlagMethodRecord corresponding to the given map/json record.
+   * @throws PiranhaConfigurationException if there was any issue reading or parsing the
+   *     configuration file.
+   */
+  static PiranhaFlagMethodRecord parseFromJSONPropertyEntryMap(
+      Map<String, Object> methodPropertyEntry, boolean isArgumentIndexOptional)
+      throws PiranhaConfigurationException {
+    String methodName = getValueStringFromMap(methodPropertyEntry, METHOD_NAME_KEY);
```

Comment:
Can code from line 111 to 132 be hoisted into a utility method and called. The same method can be called from PiranhaTestMethodRecord too. The messaging in the `throw` and the check on `flagType` seem to be the only differences between the two methods.